### PR TITLE
Middleware for rails, sinatra and sidekiq

### PR DESCRIPTION
**Pull request by igorvpcleao**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+source 'https://rubygems.org'
+ruby '2.5.1'
+
+gemspec
```
Comment (cscatolini) [inline Gemfile]: new line

File: lib/datadog-client.rb [hunk 2]
```diff
@@ -0,0 +1 @@
+require "ruby/reporters/datadog"
```
Comment (cscatolini) [inline lib/datadog-client.rb]: new line

File: lib/rails-middleware-datadog.rb [hunk 3]
```diff
@@ -0,0 +1 @@
+require "rails/middleware/datadog"
```
Comment (cscatolini) [inline lib/rails-middleware-datadog.rb]: new line (you should probably configure your editor to add it automatically)

File: lib/rails/middleware/datadog.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      def report_to_statsd(start_request_time, end_request_time, env, status)
+        elapsed_time_ms = compute_elapsed_time_ms(start_request_time, end_request_time)
+        request_method = extract_method_from_request(env)
+        path = extract_path_from_request(env)
+
+        @statsd.response_time_ms(path, request_method, status, elapsed_time_ms)
+      end
+    end
+  end
+end
```
Comment (cscatolini) [inline lib/rails/middleware/datadog.rb]: new line

File: lib/sidekiq-middleware-datadog.rb [hunk 5]
```diff
@@ -0,0 +1 @@
+require "sidekiq/middleware/datadog"
```
Comment (cscatolini) [inline lib/sidekiq-middleware-datadog.rb]: new line

File: lib/sidekiq/middleware/datadog.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
+require 'datadog/statsd'
+require 'socket'
+require 'sidekiq'
+
+module Sidekiq
+  module Middleware
+    class Datadog
+      attr_accessor :hostname, :statsd_host, :statsd_port, :statsd, :xalala
```
Comment (cscatolini) [inline lib/sidekiq/middleware/datadog.rb]: is `xalala` supposed to be here?

File: lib/sinatra-middleware-datadog.rb [hunk 7]
```diff
@@ -0,0 +1 @@
+require "sinatra/middleware/datadog"
```
Comment (cscatolini) [inline lib/sinatra-middleware-datadog.rb]: new line

File: lib/sidekiq/middleware/datadog.rb [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        queue = queue_name ? Sidekiq::Queues[queue_name] : Sidekiq::Queues["default"]
+        @statsd.gauge "#{@metric_name}.queue_size", queue.size, :tags => tags
+      end
+
+      def report_queue_processing_time(value, tags)
+        @statsd.timing "#{@metric_name}.run_time_ms", value, :tags => tags
+      end
+    end
+  end
+end
```
Comment (cscatolini) [inline lib/sidekiq/middleware/datadog.rb]: new line

File: lib/sinatra/middleware/datadog.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+require "rails/middleware/datadog"
+
+module Sinatra
+  module Middleware
+    class Datadog < Rails::Middleware::Datadog
+    end
+  end
+end
```
Comment (cscatolini) [inline lib/sinatra/middleware/datadog.rb]: new line

File: spec/rails/middleware/datadog_spec.rb [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            ]])
+        end
+      end
+    end
+  end
+
+  def env_for(url, opts={})
+    Rack::MockRequest.env_for(url, opts)
+  end
+end
```
Comment (cscatolini) [inline spec/rails/middleware/datadog_spec.rb]: new line

File: spec/ruby/reporters/datadog_spec.rb [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+          @statsd.increment("number-of-pitayas", ['kind:juice'])
+
+          expect(@statsd.statsd.socket.buffer).to eq([[
+            "number-of-pitayas:1|c|#kind:juice"
+            ]])
+        end
+      end
+    end
+  end
+end
```
Comment (cscatolini) [inline spec/ruby/reporters/datadog_spec.rb]: new line

File: spec/sidekiq/datadog_spec.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+          ])
+
+          expect(@statsd.socket.buffer[1]).to eq([
+            "worker.queue_size:1|g|#custom:tag,worker:0,env:test,name:mock/worker,queue:default,error:false"
+          ])
+        end
+      end
+    end
+  end
+end
```
Comment (cscatolini) [inline spec/sidekiq/datadog_spec.rb]: new line

File: spec/spec_helper.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  
+    def to_s
+      inspect
+    end
+  
+    def inspect
+      "<FakeUDPSocket: #{@buffer.inspect}>"
+    end
+  end
+end
```
Comment (cscatolini) [inline spec/spec_helper.rb]: new line

**Review by cscatolini — APPROVED** [review]
Fix the new lines and we're good to merge.

File: Gemfile (hunk 1 above)
Comment (igorvpcleao) [reply Gemfile]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: lib/datadog-client.rb (hunk 2 above)
Comment (igorvpcleao) [reply lib/datadog-client.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: lib/rails-middleware-datadog.rb (hunk 3 above)
Comment (igorvpcleao) [reply lib/rails-middleware-datadog.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

Comment (igorvpcleao) [reply lib/rails-middleware-datadog.rb]: configured

File: lib/rails/middleware/datadog.rb (hunk 4 above)
Comment (igorvpcleao) [reply lib/rails/middleware/datadog.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: lib/sidekiq-middleware-datadog.rb (hunk 5 above)
Comment (igorvpcleao) [reply lib/sidekiq-middleware-datadog.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: lib/sinatra/middleware/datadog.rb (hunk 9 above)
Comment (igorvpcleao) [reply lib/sinatra/middleware/datadog.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: spec/rails/middleware/datadog_spec.rb (hunk 10 above)
Comment (igorvpcleao) [reply spec/rails/middleware/datadog_spec.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: spec/sidekiq/datadog_spec.rb (hunk 12 above)
Comment (igorvpcleao) [reply spec/sidekiq/datadog_spec.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

File: spec/spec_helper.rb (hunk 13 above)
Comment (igorvpcleao) [reply spec/spec_helper.rb]: c0e0a8ca5f1798fb3298a3bbc8773fa799921a18

**Pull request merged**